python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -146,52 +216,223 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
-            run_config,
-            datasets,
-            preprocessor,
-            resume_from_checkpoint,
+            train_loop_per_worker=_huggingface_train_loop_per_worker,
+            train_loop_config=trainer_init_config,
+            torch_config=torch_config,
+            scaling_config=scaling_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
         )
 
-    def _create_train_func(self, trainer_init_per_worker):
-        def train_loop_per_worker(config):
-            # Set to None just to make CI pass & show
-            # the intended usage with trainer_init_per_worker
-            train_dataset = None
-            eval_dataset = None
-            trainer = trainer_init_per_worker(train_dataset, eval_dataset, **config)
-            trainer.train()
+    def _validate_trainer_init_per_worker(
+        self, trainer_init_per_worker: Callable, fn_name: str
+    ) -> None:
+        num_params = len(inspect.signature(trainer_init_per_worker).parameters)
+        if num_params < 3:
+            raise ValueError(
+                f"{fn_name} should take in at least 3 arguments, "
+                f"but it accepts {num_params} arguments instead."
+            )
+
+    def _validate_attributes(self):
+        # exceptions first
+        if TRAIN_DATASET_KEY not in self.datasets:
+            raise KeyError(
+                f"'{TRAIN_DATASET_KEY}' key must be preset in `datasets`. "
+                f"Got {list(self.datasets.keys())}"
+            )
+        if not all(
+            key in (TRAIN_DATASET_KEY, EVALUATION_DATASET_KEY) for key in self.datasets
+        ):
+            raise KeyError(
+                f"Only '{TRAIN_DATASET_KEY}' and '{EVALUATION_DATASET_KEY}' "
+                "keys can be preset in `datasets`. "
+                f"Got {list(self.datasets.keys())}"
+            )
+        gpus_per_worker = self.scaling_config.get("num_gpus_per_worker", 0)
+        if gpus_per_worker > 1:
+            raise ValueError(
+                f"You have assigned {gpus_per_worker} GPUs per worker. "
+                "This is not supported by HuggingFace, which expects "
+                "one GPU per worker in DDP mode and will fail "
+                "if more are assigned."
+            )
+
+        super()._validate_attributes()
+
+    def as_trainable(self) -> Type[Trainable]:
+        # Replace the directory checkpoint with a node ip & path dict checkpoint
+        # used to sync the directory. If we use a directry checkpoint directly,
+        # it will get deepcopied & serialized unnecessarily
+        original_param_dict = self._param_dict.copy()
+        resume_from_checkpoint: Optional[Checkpoint] = self._param_dict.get(
+            "resume_from_checkpoint", None
+        )
+        if resume_from_checkpoint:
+            with resume_from_checkpoint.as_directory() as checkpoint_path:
```

Comment:
Actually I'm not sure if this will work when Tune internally tries to resume from failure rather than the user passing in a checkpoint to resume from: https://github.com/ray-project/ray/blob/master/python/ray/ml/trainer.py#L347-L352.

I think we need to move this logic into trainer.setup.